google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/GradleBuildDeploymentSource.java

```diff
@@ -0,0 +1,116 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.intellij.appengine.cloud;
+
+import com.intellij.openapi.components.ServiceManager;
+import com.intellij.openapi.externalSystem.service.project.manage.ProjectDataManager;
+import com.intellij.openapi.module.ModulePointer;
+import com.intellij.openapi.project.Project;
+import com.intellij.remoteServer.configuration.deployment.DeploymentSourceType;
+import com.intellij.remoteServer.impl.configuration.deployment.ModuleDeploymentSourceImpl;
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.jetbrains.plugins.gradle.model.ExternalProject;
+import org.jetbrains.plugins.gradle.service.project.data.ExternalProjectDataService;
+import org.jetbrains.plugins.gradle.util.GradleConstants;
+
+import java.io.File;
+import java.io.FilenameFilter;
+
+import javax.swing.Icon;
+
+import icons.GradleIcons;
+
+/**
+ * A deployment source backed by the Gradle build system
+ */
+public class GradleBuildDeploymentSource extends ModuleDeploymentSourceImpl {
+
+  private final Project project;
+
+  public GradleBuildDeploymentSource(@NotNull ModulePointer pointer, @NotNull Project project) {
+    super(pointer);
+    this.project = project;
+  }
+
+  @NotNull
+  @Override
+  public String getPresentableName() {
+    return String.format("Gradle build: %s", getModulePointer().getModuleName());
+  }
+
+  @Nullable
+  @Override
+  public Icon getIcon() {
+    return GradleIcons.Gradle;
+  }
+
+  @NotNull
+  @Override
+  public DeploymentSourceType<?> getType() {
+    return DeploymentSourceType.EP_NAME.findExtension(GradleBuildDeploymentSourceType.class);
+  }
+
+  @Nullable
+  @Override
+  public File getFile() {
+    ExternalProjectDataService dataService =
+        (ExternalProjectDataService) ServiceManager.getService(
+            ProjectDataManager.class).getDataService(ExternalProjectDataService.KEY);
+
+    if (dataService == null || project.getBasePath() == null || getModule() == null) {
+      return null;
+    }
+
+    ExternalProject rootExternalProject =
+        dataService.getRootExternalProject(GradleConstants.SYSTEM_ID,
+            new File(project.getBasePath()));
+
+    if (rootExternalProject == null) {
+      return null;
+    }
+
+    ExternalProject moduleExternalProject =
+        dataService.findExternalProject(rootExternalProject, getModule());
+
+    if (moduleExternalProject == null) {
+      return null;
+    }
+
+    return findGradleBuildArtifact(moduleExternalProject.getBuildDir());
+  }
+
+  /**
+   * Uses a best-effort approach to locate the Gradle build artifact.
+   * Assumes the build artifact is located in the /libs folder under the root build directory.
+   * Also assumes only a single build artifact. Will only return one artifact if multiple are found.
+   */
+  @Nullable
+  private File findGradleBuildArtifact(@NotNull File buildDir) {
```

Comment:
aha.  I thought that we had found a robust way to do this.  So Gradle does not let you determine the build artifact for a given module?